### PR TITLE
feat(plonky2): allow writing the PLONKY2 operations to file

### PR DESCRIPTION
**Pull request by nickysn**

### CODE REVIEW

File: compiler/noirc_driver/src/lib.rs [hunk 1]
```diff
@@ -70,6 +70,9 @@ pub struct CompileOptions {
     #[arg(long, hide = true)]
     pub show_plonky2: bool,
 
+    #[arg(long, hide = true)]
+    pub write_plonky2_to_file: Option<String>,
```
Comment (smanilov) [inline compiler/noirc_driver/src/lib.rs:74]: Rename to `plonky2_print_file`? The current name sounds like the name of a bool; so anything that sounds like the name of an (optional) file would be preferable.

File: compiler/noirc_evaluator/src/ssa/plonky2_gen/console_asm_writer.rs [hunk 2]
```diff
@@ -165,6 +170,23 @@ impl Display for U32TargetDisplay {
...
+
+    fn handle_output(&mut self, s: String) {
+        if self.show_plonky2 {
+            println!("{}", s);
+        }
+
+        if let Some(f) = &mut self.file {
+            writeln!(f, "{}", s).expect("Unable to write PLONKY2 data to file");
+        }
+    }
```
Comment (smanilov) [inline compiler/noirc_evaluator/src/ssa/plonky2_gen/console_asm_writer.rs:189]: I would argue that the struct name is not so good anymore, now that it also writes to a file... :-)

How about we simplify things by passing an instance of AsmWriter to Builder::new and we don't have it parametrized by type and we don't have a separate impl and trait?

File: compiler/noirc_driver/src/lib.rs (hunk 1 above)
Comment (nickysn) [reply compiler/noirc_driver/src/lib.rs]: Done: https://github.com/blocksense-network/noir/pull/34/commits/29e1e050d57bd7cea169402f743d3a4ebba0d69d

File: compiler/noirc_evaluator/src/ssa/plonky2_gen/console_asm_writer.rs (hunk 2 above)
Comment (nickysn) [reply compiler/noirc_evaluator/src/ssa/plonky2_gen/console_asm_writer.rs]: Ok. I removed the trait AsmWriter and renamed the struct ConsoleAndFileAsmWriter to AsmWriter.

Comment (smanilov) [reply compiler/noirc_evaluator/src/ssa/plonky2_gen/console_asm_writer.rs]: Great! Thanks!

**Comment by smanilov** [comment]
I ran `cargo run --bin nargo prove --show-plonky2 > /tmp/print` and `cargo run --bin nargo prove --plonky2-print-file /tmp/print2` and verified that the two files are identical.

**Review by smanilov — APPROVED** [review]

**Comment by zah** [comment]
> I ran cargo run --bin nargo prove --show-plonky2 > /tmp/print and cargo run --bin nargo prove --plonky2-print-file /tmp/print2 and verified that the two files are identical.

It seems to me that we can have a good test suite around the PLONKY2 constraint printing functionality that will be able to verify this (among other things).

Such a suite can compare the constraints listing of specific programs against some expected output. It may turn out to be fragile (the constraint listing getting changed as we bump Noir or modify our compiler logic), but I would argue that it's better for us to be aware of such changes instead of allowing silent regression in our code generation quality.

**Comment by smanilov** [comment]
I think being able to write tests was a major motivation for Niki to submit this PR, so we agree and we're on it :-)

**Pull request merged**